### PR TITLE
Fix account switching and chat history issue

### PR DESCRIPTION
**Pull request by mkondratek**
Fixes https://github.com/sourcegraph/jetbrains/issues/2173.

## Test plan
See the issue: https://github.com/sourcegraph/jetbrains/issues/2173

1. log in to an account 
2. write something in chat
3. switch account
4. review the chat history
expected:
The chat from the first account should not appear in the account switched into.

### CODE REVIEW

File: vscode/src/chat/chat-view/ChatsController.ts [hunk 1]
```diff
@@ -500,6 +494,7 @@ export class ChatsController implements vscode.Disposable {
             extensionClient: this.extensionClient,
         })
     }
+    play
```
Comment (valerybugakov) [inline vscode/src/chat/chat-view/ChatsController.ts:497]: Redundant? 
```suggestion
```

File: vscode/src/chat/chat-view/ChatController.ts [hunk 2]
```diff
@@ -1508,9 +1509,9 @@ export class ChatController implements vscode.Disposable, vscode.WebviewViewProv
         this.postViewTranscript()
     }
 
-    private async saveSession(): Promise<void> {
+    private async saveSession(authStatus?: AuthStatus): Promise<void> {
         const allHistory = await chatHistory.saveChat(
-            this.authProvider.getAuthStatus(),
+            authStatus ? authStatus : this.authProvider.getAuthStatus(),
```
Comment (valerybugakov) [inline vscode/src/chat/chat-view/ChatController.ts:1514]:
```suggestion
            authStatus || this.authProvider.getAuthStatus(),
```

File: vscode/src/chat/chat-view/ChatController.ts [hunk 3]
```diff
@@ -588,14 +588,15 @@ export class ChatController implements vscode.Disposable, vscode.WebviewViewProv
     // #region top-level view action handlers
     // =======================================================================
 
-    public setAuthStatus(status: AuthStatus): void {
+    public async setAuthStatus(status: AuthStatus): Promise<void> {
         // Run this async because this method may be called during initialization
         // and awaiting on this.postMessage may result in a deadlock
         void this.sendConfig()
```
Comment (valerybugakov) [inline vscode/src/chat/chat-view/ChatController.ts:594]: Should we worry about a deadlock mentioned in the comment while we add `await` to this function?

**Review by valerybugakov — APPROVED** [review]
Verified that the fix works based on the test plan. 

Would it be more resilient to keep one source of truth for the current `authStatus` by disposing all chats _before_ changing the value of `this.authProvider.getAuthStatus()`? If we do all the cleanup tasks before setting the new value, all the code paths can still rely exclusively on `this.authProvider.getAuthStatus()`, which makes the implementation more robust. 

I might be missing something obvious here, so I approve the fix.

File: vscode/src/chat/chat-view/ChatController.ts (hunk 3 above)
Comment (mkondratek) [reply vscode/src/chat/chat-view/ChatController.ts:594]: I am not 100% sure about it but I think the comment is about `void this.sendConfig()` call.

Comment (mkondratek) [reply vscode/src/chat/chat-view/ChatController.ts:594]: Looks like some old leftovers: https://github.com/sourcegraph/cody/blob/eabbdcf735592796438e44ba1fdcb3777f5098c3/vscode/src/chat/chat-view/SimpleChatPanelProvider.ts#L518

Comment (mkondratek) [reply vscode/src/chat/chat-view/ChatController.ts:594]: The comment applies to `postMessage({type: 'config', ...)` - I do not change that. I uphold that my changes should not break anything (at least not in this area 😅)

**Comment by mkondratek** [comment]
> Verified that the fix works based on the test plan.
> 
> Would it be more resilient to keep one source of truth for the current `authStatus` by disposing all chats _before_ changing the value of `this.authProvider.getAuthStatus()`? If we do all the cleanup tasks before setting the new value, all the code paths can still rely exclusively on `this.authProvider.getAuthStatus()`, which makes the implementation more robust.
> 
> I might be missing something obvious here, so I approve the fix.

I agree with you. Presently, `ChatsController.setAuthStatus` does two things - unsets the old auth status (disposes chats) and sets the new auth status (loads chats). It looks like we should split these two operations somehow. However, I do not feel like having enough proficiency in TS to apply the ideal solution. 

Also note that some of our problem are going to be solved once we switch to webview account management (which is our plan afaik).

**Comment by mkondratek** [comment]
Thanks for your input and review @valerybugakov !

**Review by pkukielka — APPROVED** [review]
LGTM

**Pull request merged**